clippy_lints/src/mixed_read_write_in_expression.rs

```diff
@@ -136,6 +137,21 @@ impl<'a, 'tcx> DivergenceVisitor<'a, 'tcx> {
 impl<'a, 'tcx> Visitor<'tcx> for DivergenceVisitor<'a, 'tcx> {
     fn visit_expr(&mut self, e: &'tcx Expr<'_>) {
         match e.kind {
+            // fix #10776
+            ExprKind::Block(block, ..) => {
+                if let Some(e) = block.expr && block.stmts.is_empty() {
+                    self.visit_expr(e);
+
+                    return;
+                }
+
+                if let [stmt, rest @ ..] = block.stmts && rest.is_empty() {
```

Comment:
This ends up linting a block like `{ diverge(); 5 }` which it used to not do.

You can match on multiple fields at the same time which would make this clearer. e.g.
```rust
match (stmts, expr) {
    ([], Some(e)) => {}
    ([s], None) => {}
    _ => {}
}
```

Reply:
> This ends up linting a block like { diverge(); 5 } which it used to not do.

I'm not sure if this is an issue as it's unreachable anyway

I'll change it to that

Reply:
Done, also added a test for that (`{ diverge(); 5 }`) and it's fixed